### PR TITLE
Harvest Billing Widget

### PR DESCRIPTION
**Pull request by BogdanBrat**
Harvest Billing widget Project

### CODE REVIEW

File: Harvest Billing Widget/widget.html [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+                            url: harvestApiUrl,
+                            method: 'GET',
+                            headers: {
+                                'Authorization': `Bearer ${authToken}`,
+                                'Harvest-Account-Id': harvestAccountId,
+                                'User-Agent': 'YourApp (yourname@example.com)'
+                            },
+                            success: function (data) {
+                                data.time_entries.forEach(function (entry) {
+                                    if (entry.project && entry.project.code && entry.project.code.includes("R&D")) {
```
Comment (BenGWeeks) [inline Harvest Billing Widget/widget.html]: As discussed, please don't hard code the filter "R&D", this should be a widget property.

File: Harvest Billing Widget/docs/security-settings.png
Comment (BenGWeeks) [inline Harvest Billing Widget/docs/security-settings.png:1]: Remove screenshots (e.g. widget-configuration and security-settings - just succinct descriptions is fine) - UI's change over time and more difficult to keep up to date.

File: Harvest Billing Widget/docs/widget-configuration.png
Comment (BenGWeeks) [inline Harvest Billing Widget/docs/widget-configuration.png:1]: Remove screenshots (e.g. widget-configuration and security-settings - just succinct descriptions is fine) - UI's change over time and more difficult to keep up to date.

**Review by BenGWeeks — CHANGES_REQUESTED** [review]
Love the new logo. Which image is used in the marketplace? logo.png or preview.png? I note that preview.png does not have a transparent background. ref: https://marketplace.visualstudio.com/).

Updating the PR description with an overview of the functionality you are submitting and a screenshot always goes down well :-)

I'm assuming the below are the default DevOps colours, so we should ensure that the colours in the widget are configurable, so that we can set those colours to our own branding.

image

Please see further specific review items.

**Comment by BenGWeeks** [comment]
Is it possible to add descriptions to some of the widget properties? Such as these (and put description of how to get the Harvest ID, etc.).

image

File: Harvest Billing Widget/README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+1. **Log in to Harvest**: Access your Harvest account.
+2. **Navigate to Profile**: Click your name in the upper right corner and select "My Profile".
+3. **Access Security Settings**: Go to "Settings" -> "Security" -> "Go to Harvest ID security settings" -> "Developers".
+4. **Create or Use Token**: Create a new personal access token or use an existing one. Copy the token and account ID and store them securely.
+
+![Security Settings](https://raw.githubusercontent.com/BogdanBrat/DevOpsWidgetHarvestHours/main/docs/security-settings.png)
+
+## Installation and Setup
+
+1. **Install the Widget**: Download and install from the Azure DevOps Marketplace.
```
Comment (BenGWeeks) [inline Harvest Billing Widget/README.md]: "Download"? If this isn't publicly available, how does a browser of this repo deploy the code to their own tenant marketplace?

**Comment by BenGWeeks** [comment]
Optional nice to have:

- Ability to rename the labels (e.g. I might prefer to rename "R&D" to "Innovation", or "Billable" to "Chargeable")
- Ability to define label colours (see image below example from another widget)

image

**Comment by BogdanBrat** [comment]
Please find this widget here:
https://marketplace.visualstudio.com/items?itemName=EirEvo-EvoLabs.HarvestWidgets

File: Harvest Billing Widget/vss-extension.json [hunk 3]
```diff
@@ -0,0 +1,100 @@
+{
+  "manifestVersion": 1,
+  "id": "HarvestWidgets",
+  "version": "1.0.7",
+  "name": "Harvest Billing Widget",
+  "description": "Displays hours from Harvest, providing detailed insights into time allocation for projects. The hours are categorized into billable hours, 'R&D' (Research & Development), and Internal hours.",
+  "publisher": "EirEvo-AzureDevOpsExtensions",
```
Comment (BenGWeeks) [inline Harvest Billing Widget/vss-extension.json]: Publisher should be "Eir Evo". I see this appears here:

image

File: Harvest Billing Widget/vss-extension.json [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    "harvest",
+    "time tracking",
+    "dashboard",
+    "widget",
+    "Internal",
+    "extension"
+  ],
+  "repository": {
+    "type": "git",
+    "uri": "https://github.com/BogdanBrat/DevOpsWidgetHarvestHours"
```
Comment (BenGWeeks) [inline Harvest Billing Widget/vss-extension.json]: This needs to be their Eir Evo GitHub repo not your personal fork.

**Comment by BogdanBrat** [comment]
For the logo in the Marketplace, I used an icon without a background because it looks more polished that way. However, for the preview in DevOps dashboards, I included a background, as it enhances the visual appeal in that context.

image

File: Harvest Billing Widget/widget.html (hunk 1 above)
Comment (BogdanBrat) [reply Harvest Billing Widget/widget.html]: Let me clarify, so we can show R&D or not depending on whether we choose it in the widget settings?

**Comment by BenGWeeks** [comment]
> For the logo in the Marketplace, I used an icon without a background because it looks more polished that way. However, for the preview in DevOps dashboards, I included a background, as it enhances the visual appeal in that context.
> 
> image

Looks good!

File: Harvest Billing Widget/widget.html (hunk 1 above)
Comment (BenGWeeks) [reply Harvest Billing Widget/widget.html]: Different businesses may set their own project codes to specify "R&D" projects. For example, we said in the project code in Harvest, we will post-fix the project code with "R&D", but this should be configurable and not hard coded.

File: Harvest Billing Widget/docs/overview.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+- **Widget Title**: Customize the title displayed on the widget to suit your needs.
+- **Widget Size**: Adjust the size of the widget to fit your dashboard layout.
+
+![Widget Configuration](https://raw.githubusercontent.com/BogdanBrat/DevOpsWidgetHarvestHours/main/docs/widget-configuration.png)
+
+### Obtaining Harvest Account ID and Authorization Token
+
+To get the necessary credentials:
+
+1. **Step 1**: Log in to your Harvest account.
```
Comment (BenGWeeks) [inline Harvest Billing Widget/docs/overview.md]: Remove "Step 1" etc. here, it's obvious from the numbered list item.

image

File: Harvest Billing Widget/docs/overview.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+2. **Step 2**: Go to your profile by clicking your name in the upper right corner and select "My Profile".
+3. **Step 3**: Navigate to "Settings" -> "Security" -> "Go to Harvest ID security settings" -> "Developers".
```
Comment (BenGWeeks) [inline Harvest Billing Widget/docs/overview.md]: Use the following format:

```
3. Navigate to `Settings` > `Security` > `Go to Harvest ID security settings` > `Developers`.
```

File: Harvest Billing Widget/docs/overview.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+4. **Step 4**: Click "Create new personal access token" or use an existing token. Copy the generated token and account ID and store them securely.
```
Comment (BenGWeeks) [inline Harvest Billing Widget/docs/overview.md]: Use the following format:

```
`Create new personal access token`
```

File: Harvest Billing Widget/docs/overview.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+1. **Install the Widget**: Download and install the Harvest Billing Widget from the Azure DevOps Marketplace.
+2. **Add to Dashboard**: Drag and drop the widget onto your desired dashboard.
+3. **Enter Configuration Details**: Provide your Harvest Account ID and Authorization Token to enable data fetching.
+4. **Customize Display Options**: Select your preferred display mode, widget title, and widget size to visualize the data effectively.
+
+## Usage
+
+After configuration, the widget provides a real-time breakdown of hours spent on different types of work. This allows teams and managers to monitor resource utilization and focus, ensuring that project hours are appropriately categorized and tracked.
+
+For further assistance or detailed documentation, please visit [Harvest Billing Widget Documentation](https://github.com/BogdanBrat/DevOpsWidgetHarvestHours).
```
Comment (BenGWeeks) [inline Harvest Billing Widget/docs/overview.md]: Link to this repo - not personal repos.

File: Harvest Billing Widget/README.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+- **Widget Title**: Customize the title displayed on the widget.
+- **Widget Size**: Adjust the widget size to fit your dashboard layout.
+
+![Widget Configuration](https://raw.githubusercontent.com/BogdanBrat/DevOpsWidgetHarvestHours/main/docs/widget-configuration.png)
+
+### Obtaining Harvest Account ID and Authorization Token
+
+1. **Log in to Harvest**: Access your Harvest account.
+2. **Navigate to Profile**: Click your name in the upper right corner and select "My Profile".
+3. **Access Security Settings**: Go to "Settings" -> "Security" -> "Go to Harvest ID security settings" -> "Developers".
```
Comment (BenGWeeks) [inline Harvest Billing Widget/README.md]: Use the format:

```
Go to `Settings` > `Security` > `Go to Harvest ID security settings` > `Developers`.
```

File: Harvest Billing Widget/README.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+- **Display Mode**: Choose to display data either as a percentage or in absolute hours.
+- **Widget Title**: Customize the title displayed on the widget.
+- **Widget Size**: Adjust the widget size to fit your dashboard layout.
+
+![Widget Configuration](https://raw.githubusercontent.com/BogdanBrat/DevOpsWidgetHarvestHours/main/docs/widget-configuration.png)
+
+### Obtaining Harvest Account ID and Authorization Token
+
+1. **Log in to Harvest**: Access your Harvest account.
+2. **Navigate to Profile**: Click your name in the upper right corner and select "My Profile".
```
Comment (BenGWeeks) [inline Harvest Billing Widget/README.md]: Use the format:

```
Click your name in the upper right corner and select `My Profile`.
```

File: Harvest Billing Widget/vss-extension.json (hunk 4 above)
Comment (BenGWeeks) [reply Harvest Billing Widget/vss-extension.json]: https://github.com/EirEvo/AzureDevOpsExtensions

File: Harvest Billing Widget/docs/overview.md (hunk 8 above)
Comment (BenGWeeks) [reply Harvest Billing Widget/docs/overview.md]: https://github.com/EirEvo/AzureDevOpsExtensions

**Comment by BogdanBrat** [comment]
> Is it possible to add descriptions to some of the widget properties? Such as these (and put description of how to get the Harvest ID, etc.).
> 
> image


**This is done.**
I created an informational description of all the parameters for the Authorization Token and Harvest Account ID fields, including a description of where to find these values.

image

**Comment by BogdanBrat** [comment]
> Optional nice to have:
> 
> * Ability to rename the labels (e.g. I might prefer to rename "R&D" to "Innovation", or "Billable" to "Chargeable")
> * Ability to define label colours (see image below example from another widget)
> 
> image

**Comment by BogdanBrat** [comment]
> Optional nice to have:
> 
> * Ability to rename the labels (e.g. I might prefer to rename "R&D" to "Innovation", or "Billable" to "Chargeable")
> * Ability to define label colours (see image below example from another widget)
> 
> image



**This is done.**
New options for choosing the color of the pie chart sectors in the widget settings have been created, along with fields for changing the names of these sectors.

image

File: Harvest Billing Widget/widget.html (hunk 1 above)
Comment (BogdanBrat) [reply Harvest Billing Widget/widget.html]: > As discussed, please don't hard code the filter "R&D", this should be a widget property.

**It's done.**
A new R&D Filter property widget was created, allowing you to enter a word to filter and select projects.

image

**Comment by BenGWeeks** [comment]
@BogdanBrat Over what period are the harvest hours calculated?

File: Harvest Billing Widget/widget.html [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+                                        rAndDHours += entry.hours;
+                                    } else if (entry.billable) {
+                                        billableHours += entry.hours;
+                                    } else {
+                                        nonBillableHours += entry.hours;
+                                    }
+                                });
+
+                                var totalHours = billableHours + nonBillableHours + rAndDHours;
+                                var ctx = document.getElementById('hoursChart').getContext('2d');
```
Comment (BenGWeeks) [inline Harvest Billing Widget/widget.html:116]: What is `2d`? Looks like something is hardcoded here.

**Comment by BogdanBrat** [comment]
> @BogdanBrat Over what period are the harvest hours calculated?

The period for calculating Harvest hours is the current month.

File: Harvest Billing Widget/widget.html [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+                        document.querySelector('.title').textContent = widgetTitle;
+
+                        var billableHours = 0;
+                        var nonBillableHours = 0;
+                        var rAndDHours = 0;
+
+                        var currentDate = new Date();
+                        var startDate = new Date(currentDate.getFullYear(), currentDate.getMonth(), 1).toISOString().split('T')[0];
+                        var endDate = new Date(currentDate.getFullYear(), currentDate.getMonth() + 1, 0).toISOString().split('T')[0];
+
```
Comment (BenGWeeks) [inline Harvest Billing Widget/widget.html:86]: Ah, this is it, it's hard coded to get this current months billability.

I think the best solution is to add "7-days", "14-days", "30-days" that the user can click on to set the period to calculate, and default to 7-days.

**Comment by BenGWeeks** [comment]
> > @BogdanBrat Over what period are the harvest hours calculated?
> 
> The period for calculating Harvest hours is the current month.

As discussed, I think the best solution is to add "7-days", "14-days", "30-days" that the user can click on to set the period to calculate, and default to 7-days

**Comment by BogdanBrat** [comment]
> > > @BogdanBrat Over what period are the harvest hours calculated?
> > 
> > 
> > The period for calculating Harvest hours is the current month.
> 
> As discussed, I think the best solution is to add "7-days", "14-days", "30-days" that the user can click on to set the period to calculate, and default to 7-days

**It's done.**
New functionality has been created for widget settings: 'Select Time Period,' with a default set to 7 days.

image
image

**Review by BenGWeeks — CHANGES_REQUESTED** [review]
You may have resolved some of these now.

File: Harvest Billing Widget/README.md (hunk 2 above)
Comment (BogdanBrat) [reply Harvest Billing Widget/README.md]: The README file has been completely updated, and there is no longer any mention of downloading the extension.

File: Harvest Billing Widget/vss-extension.json (hunk 3 above)
Comment (BogdanBrat) [reply Harvest Billing Widget/vss-extension.json]: A new 'Eir Evo - Evo Labs' publisher, has been created to handle the publication of our upcoming extensions.
image

File: Harvest Billing Widget/docs/overview.md (hunk 5 above)
Comment (BogdanBrat) [reply Harvest Billing Widget/docs/overview.md]: To determine which key phrases will be searched for in the project field in Harvest, input text has been configured in the extension settings.

image

The overview file has been updated.

image

File: Harvest Billing Widget/docs/overview.md (hunk 6 above)
Comment (BogdanBrat) [reply Harvest Billing Widget/docs/overview.md]: The overview file has been updated.

image

File: Harvest Billing Widget/docs/overview.md (hunk 7 above)
Comment (BogdanBrat) [reply Harvest Billing Widget/docs/overview.md]: image

File: Harvest Billing Widget/docs/overview.md (hunk 8 above)
Comment (BogdanBrat) [reply Harvest Billing Widget/docs/overview.md]: The documentation link has been changed.

image

File: Harvest Billing Widget/README.md (hunk 9 above)
Comment (BogdanBrat) [reply Harvest Billing Widget/README.md]: The README file has been completely updated, and there is no longer any mention this text.

File: Harvest Billing Widget/README.md (hunk 10 above)
Comment (BogdanBrat) [reply Harvest Billing Widget/README.md]: image

File: Harvest Billing Widget/widget.html (hunk 12 above)
Comment (BogdanBrat) [reply Harvest Billing Widget/widget.html:86]: A new logic for selecting the period to display data from Harvest has been implemented.

image

File: Harvest Billing Widget/widget.html (hunk 11 above)
Comment (BogdanBrat) [reply Harvest Billing Widget/widget.html:116]: '2d' refers to a two-dimensional drawing context for an HTML canvas, allowing the creation and rendering of graphical elements such as lines, shapes, and images on a plane with X and Y coordinates.

File: Harvest Billing Widget/vss-extension.json (hunk 4 above)
Comment (BogdanBrat) [reply Harvest Billing Widget/vss-extension.json]: File vss-extension.json has been updated.

image

**Comment by BogdanBrat** [comment]
@BenGWeeks , please review my comments and provide feedback to confirm if everything meets our expectations.

**Pull request merged**